### ISSUE TITLE
mqtt-sn Gateway not working

### ISSUE THREAD

**Issue by pritstift** [emqtt/emq-sn#12]
I played around with the emq_sn plugin. I am trying to publish some messages using [mqtt-sn-tools](https://github.com/njh/mqtt-sn-tools). 

`./mqtt-sn-pub -d -d -d -h 127.0.0.1 -p 1884 -q 0 -r  -t world -m hello `

However as the below output shows, i do not receive  meaningful topic id:

```
2017-04-11 13:37:25 DEBUG Debug level is: 3.
2017-04-11 13:37:25 DEBUG Connecting...
2017-04-11 13:37:25 DEBUG Sending CONNECT packet...
2017-04-11 13:37:25 DEBUG Sending  24 bytes. Type=CONNECT on Socket: 3.
2017-04-11 13:37:25 DEBUG waiting for packet...
2017-04-11 13:37:25 DEBUG Received  3 bytes from 127.0.0.1:1884. Type=CONNACK on Socket: 3
2017-04-11 13:37:25 DEBUG CONNACK return code: 0x00
2017-04-11 13:37:25 DEBUG Sending REGISTER packet...
2017-04-11 13:37:25 DEBUG Sending  11 bytes. Type=REGISTER on Socket: 3.
2017-04-11 13:37:25 DEBUG waiting for packet...
2017-04-11 13:37:25 DEBUG Received  7 bytes from 127.0.0.1:1884. Type=REGACK on Socket: 3
2017-04-11 13:37:25 DEBUG REGACK return code: 0x00
2017-04-11 13:37:25 DEBUG REGACK topic id: 0x0000
2017-04-11 13:37:25 DEBUG Sending PUBLISH packet...
2017-04-11 13:37:25 DEBUG Sending  12 bytes. Type=PUBLISH on Socket: 3.
2017-04-11 13:37:25 DEBUG Disconnecting...
2017-04-11 13:37:25 DEBUG Sending DISCONNECT packet...
2017-04-11 13:37:25 DEBUG Sending   2 bytes. Type=DISCONNECT on Socket: 3.
```

Subscribing to a topic does also not work for me. I get a topic id that is unequal to 0x0000. So far so good. But no packages arrive at the subscriber:

`mqtt-sn-sub -d -d -d -h 127.0.0.1 -p 1884  -V -t world`

```
2017-04-11 13:44:47 DEBUG Verbose level is: 2.
2017-04-11 13:44:47 DEBUG Network timeout is: 5 seconds.
2017-04-11 13:44:47 DEBUG Connecting...
2017-04-11 13:44:47 DEBUG Sending CONNECT packet...
2017-04-11 13:44:47 DEBUG Sending  24 bytes. Type=CONNECT on Socket: 3.
2017-04-11 13:44:47 DEBUG waiting for packet...
2017-04-11 13:44:47 DEBUG Received  3 bytes from 127.0.0.1:1884. Type=CONNACK on Socket: 3
2017-04-11 13:44:47 DEBUG CONNACK return code: 0x00
2017-04-11 13:44:47 DEBUG Sending SUBSCRIBE packet...
2017-04-11 13:44:47 DEBUG Sending  10 bytes. Type=SUBSCRIBE on Socket: 3.
2017-04-11 13:44:47 DEBUG waiting for packet...
2017-04-11 13:44:47 DEBUG Received  8 bytes from 127.0.0.1:1884. Type=SUBACK on Socket: 3
2017-04-11 13:44:47 DEBUG SUBACK return code: 0x00
2017-04-11 13:44:47 DEBUG SUBACK topic id: 0x0001
2017-04-11 13:44:47 DEBUG Registering topic 0x0001: world
2017-04-11 13:44:57 DEBUG Sending PINGREQ packet...
2017-04-11 13:44:57 DEBUG Sending   2 bytes. Type=PINGREQ on Socket: 3.
2017-04-11 13:44:57 DEBUG waiting for packet...
2017-04-11 13:44:57 DEBUG Received  2 bytes from 127.0.0.1:1884. Type=PINGRESP on Socket: 3
2017-04-11 13:45:07 DEBUG Sending PINGREQ packet...
2017-04-11 13:45:07 DEBUG Sending   2 bytes. Type=PINGREQ on Socket: 3.
2017-04-11 13:45:17 ERROR Keep alive error: timed out while waiting for a PUBLISH from gateway.
```

Even though i can see the client in the webinterface of the EMQ broker with the correct timeout set to 10, the client times out after the first successful PINGREQ.

Any idea?

**Comment by grutabow** [comment]
@pritstift 

mqtt-sn-pub has bug which use topicIdType=0 in publish message. Only topicIdType 1 or 2 is allowedin publish message. To get round this issue, you may use 'short topic'(2 byte topic), for example:
    
 ./mqtt-sn-pub -d -d -d -h 127.0.0.1 -p 1884 -q 0 -r -t Hi -m world


For the mqtt-sn subscribe issue, a bug is found. You may try the latest emq-sn develop branch, or wait for next official emqtt release.

You may try my fork of mqtt-sn-tools, which has fixed bugs mentioned above:
https://github.com/grutabow/mqtt-sn-tools

**Comment by pritstift** [comment]
Okay. I fiddled a bit around with Version 2.1.1 and 2.1.2. Regarding the subscription everything seems to work.

I'm still having problems with publishing. Reading the [MQTT-SN specification v.1.2](http://mqtt.org/new/wp-content/uploads/2009/06/MQTT-SN_spec_v1.2.pdf)  i found:

> TopicIdType: indicates whether the field TopicId or TopicName included in this message contains a normal topic id (set to “0b00”), a pre-defined topic id (set to “0b01”), or a short topic name (set to “0b10”). The value “0b11” is reserved. Refer to sections 3 and 6.7 for the definition of the various types of topic ids.

Only for QoS = -1, the restriction on topicIdType = {1,2} is actually imposed, as far as i know.

Therefore, I don't think that TopicIdType = 0x00 is a bug in mqtt-sn-pub, when the topic is something long (i.e. more than 2 bytes) and qos ≠ -1. Please correct me if i am wrong at this point.

Furthermore EMQ is printing som errors when i publish a message with mqtt-sn (i don't know whether that is important for this topic):
```
16:34:23.218 [error] MQTT-SN(127.0.0.1:55355): UNEXPECTED INFO: emit_stats
16:34:23.218 [error] MQTT-SN(127.0.0.1:55355): UNEXPECTED INFO: {subscribe,[{<<"$client/mqtt-sn-tools-62490">>,1}]}
```

**Comment by grutabow** [comment]
Regarding MQTT-SN specification v.1.2:

In "5.3.4 Flags", present a general description of TopidIdType:
> TopicIdType: indicates whether the field TopicId or TopicName included in this `message` contains a normal topic id (set to “0b00”), a pre-defined topic id (set to “0b01”), or a short topic name (set to “0b10”). The value “0b11” is reserved. Refer to sections 3 and 6.7 for the definition of the various types of topic ids.

In "5.4.15 SUBSCRIBE", following rule is defined:
> TopicIdType: indicates the type of information included at the end of the message, namely “0b00”topic name, “0b01” pre-defined topic id, “0b10” short topic name, and “0b11” reserved.

In "5.4.12 PUBLISH", following rule is defined:
> TopicIdType: indicates the type of the topic id contained in the TopicId field. 
> TopicId: contains the topic id value or the short topic name for which the data is published.

That is to say:
- SUBSCRIBE allows 3 types of topic: normal topic, pre-defined topic and short topic.
- PUBLISH allows 2 types of topic: pre-defined topic and short topic.

As you know, PUBLISH is too short to held a normal topic which may be tens bytes long. As a tradeoff, MQTT-SN use a TopicId to represent such a long topic.

Till now, emq-sn does not support Qos -1 which may be implemented later.

These two 'UNEXPECTED INFO' are advanced features of emqttd, and emq-sn does not support them right now. Please ignore them which has no relation with basic subscribe/publish function.

Please don't be hesitate to report any problems you found.

**Comment by pritstift** [comment]
Ok that clears some things up. Thanks!

**Comment by jpmzometa** [comment]
Hi, I also have this problem. 

All of the MQTT-SN implementation that I have tried so far interpret the specification so that PUBLISH accept 3 topic ID types:  normal topic name, pre-defined topic id and short topic name.  That is how RSMB, mosquitto, and the official eclipse/paho mqtt-sn libraries (originally written at IBM) work.

Would you please reconsider your interpretation of the specification, and implement the gateway in a manner that is compatible with all other available tools?  Otherwise it will create headaches for people trying EMQ for the first time.

**Comment by grutabow** [comment]
@jpmzometa   Next official release will be compatible with "normal topic id" in publish message.

**Comment by grutabow** [comment]
This change will be released in version 2.2-rc.1.
